PP-790: qmgr output has inconsistent whitespaces

agrawalravi90 opened the pull request:
#### Issue-ID
<!--- Replace XXXX below (2 places) with the actual JIRA id -->
* **[PP-790](https://pbspro.atlassian.net/browse/PP-790)**
* **[EDD](https://pbspro.atlassian.net/wiki/display/PD/PP-790%3A+qmgr+output+has+inconsistent+whitespaces)**
* **[Forum Discussion](http://community.pbspro.org/t/pp-790-qmgr-output-has-inconsistent-whitespaces/566)**


#### Problem description
* qmgr output has inconsistent whitespaces

#### Cause / Analysis
* For all commands (qstat, pbsnodes etc.), the code uses spaces (" ") as prefix for new attributes, but for qmgr it uses tabs. Similarly, for line extensions in qmgr, it uses spaces as prefix while the other commands use tabs. This ticket was basically intended to make qmgr's output consistent with other commands.

#### Solution description
* Modified the code inside display() to add 4 spaces as prefix for new attributes and tabs as prefix for line extensions, which is what qstat and pbsnodes use.
* Also added a new test suite TestQmgr and added a PTL test to check for whitespace consistency.

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [x] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

prakashcv13 commented:
Ravi,

Wont we need external interface change document for this?

Thanks,
Prakash

agrawalravi90 commented:
I was in two minds about it, but if you think that this needs an interface change document as well then I'll create one. Thanks.

agrawalravi90 commented:
Ok, I just started a forum discussion and created the EDD, please see description of the PR for links. Thanks.

agrawalravi90 commented:
Requesting reviews, this is a pretty small change, thanks a bunch!

arungrover commented on test/tests/functional/pbs_qmgr.py:
```diff
@@ -0,0 +1,128 @@
+# trademark licensing policies.
+
+import os
+from string import ascii_letters
```
I do not see these two modules being used in this ptl script

agrawalravi90 replied:
I use os to remove the temporary file created, but you are right about ascii_letters, I'll remove it.

agrawalravi90 commented on test/tests/functional/pbs_qmgr.py:
```diff
@@ -0,0 +1,128 @@
+        Check that the prefix for new attributes listed out by qmgr list
+        are spaces and that for line extensions is a tab
+        """
+        pass
```
oh wow, that's a bug, that explains why the test would pass with 'check_whitespace_prefix' instead of '__check_whitespace_prefix', thanks for catching this.

agrawalravi90 commented on test/tests/functional/pbs_qmgr.py:
```diff
@@ -0,0 +1,128 @@
+        with open(fn, "w+") as tempfd:
+            ret = self.du.run_cmd(self.server.hostname, qmgr_cmd,
+                                  stdout=tempfd)
+            self.assertTrue(ret['rc'] == 0)
```
hmm, thanks for pointing that out, it will not get handled. I'll wrap the code inside a try-finally block to make sure that cleanup happens regardless of exceptions.

agrawalravi90 commented on test/tests/functional/pbs_qmgr.py:
```diff
@@ -0,0 +1,128 @@
+        # Remove the temporary file
+        os.remove(fn)
+        # Delete the vnode created
+        self.mom.delete_vnode_defs(vndef_file)
```
Actually this would leave the vnode as it as when I'd run this test. Maybe it's because this suite has only one test case so setup() doesn't get called again to clean stuff up.

agrawalravi90 commented:
I've addressed your comments Arun, please provide further feedback, thanks!

arungrover commented on test/tests/functional/pbs_qmgr.py:
```diff
@@ -37,10 +37,10 @@
 
 import os
-from string import ascii_letters
 from random import choice
```
In my previous comment I meant to point 'from random import choice' as the second redundant import :)

arungrover commented on test/tests/functional/pbs_qmgr.py:
```diff
@@ -72,57 +72,56 @@ def test_listcmd_whitespaces(self):
+                qmgr_cmd = ["qmgr", "-c", "list node " + nodename]
+            else:
+                qmgr_cmd = ["qmgr", "-c", "\'list node " + nodename + "\'"]
+            tempfd = open(fn, "w+")
```
just curious why you are opening file in w+ mode here but in write mode in the previous open call.

agrawalravi90 replied:
They should both be "w+" as we write and read both, thanks for catching this.

agrawalravi90 commented:
@arungrover I addressed your comment, also fixed some bugs that I had introduced with my previous commit (sorry that i forgot to test them), please let me know if things look okay now. Thanks!

arungrover commented on test/tests/functional/pbs_qmgr.py:
```diff
@@ -98,11 +100,19 @@ def test_listcmd_whitespaces(self):
             long_string += blah
             # Create a new vnode
             attrs = {ATTR_rescavail: 2}
```
I missed to add a comment here in last round. ATTR_rescavail is just "resoures_available" you do need ".<resource_name>" appended to it to make it meaningful. I am a little surprised to know that pbs_mom is conveniently ignoring such abnormality in vnodedef file.

arungrover commented on test/tests/functional/pbs_qmgr.py:
```diff
@@ -98,11 +100,19 @@ def test_listcmd_whitespaces(self):
             # Create a new vnode
             attrs = {ATTR_rescavail: 2}
-            vndef = self.mom.create_vnode_def("vn", attrs)
+            vndef = self.mom.create_vnode_def(node_prefix, attrs)
```
If the purpose here is just to add one vnode then please use self.server.create_vnodes(nodename, attrs, 1, self.mom). This will make sure that vnode is created with attributes you specified.
If you do this then you don't have to check for mom to be up

agrawalravi90 commented:
Thanks Arun, I addressed your comments but, I apologize, I accidentally did a git --amend on the previous commit, so the new changes got merged with the old ones, I hope it won't be too much of a pain to review them. Please let me know if things look okay now.

arungrover commented on test/tests/functional/pbs_qmgr.py:
```diff
@@ -37,7 +37,7 @@
+import time
 
 from tests.functional import *
 from ptl.lib.pbs_ifl_mock import *
```
This import is not needed. it should get imported with tests.functional.

agrawalravi90 commented:
Thanks Arun, I've removed the import and squashed the commits, requesting @PBSPro/pbspro-maintainers to merge this in if all looks good. Thanks!

arungrover commented:
Thanks for making changes @agrawalravi90. Please post the result of latest PTL script on the ticket. Let's wait for one more review and then it's ready for merge.

agrawalravi90 commented:
@arungrover sorry for the delay, I've attached the latest logs.

@PBSPro/pbspro-maintainers Arun has already signed-off on this, can one of you please review this and merge it in? It's a pretty small change. Thanks!

bhroam commented on src/cmds/qmgr.c:
```diff
@@ -2211,25 +2211,29 @@ display(int otype, int ptype, char *oname, struct batch_status *status,
 						if (rtm) {
-							printf("\ttype = %s\n", rtm->rtm_rname);
+							printf("%*s", indent_len, " ");
+							printf("type = %s\n", rtm->rtm_rname);
```
Did you have a reason to make these two printf()'s instead of one?  It's just one line of output

agrawalravi90 replied:
it was getting a bit confusing because of the "%*s" as that single formatter requires two arguments (the length and the blank space). So, i thought it'd be more clear if I separated it out. But if you think it doesn't make much difference, I'll combine them into one.

bhroam replied:
I wouldn't find it confusing, but if you do, let's err on the side of readability and leave it as twl.

bhroam commented on test/tests/functional/pbs_qmgr.py:
```diff
@@ -0,0 +1,126 @@
+            return
+
+        if line[0] == " ":
+            # Spaces are prefix for new attribute lines
```
I think you need the word 'the' in front of prefix.

bhroam commented on test/tests/functional/pbs_qmgr.py:
```diff
@@ -0,0 +1,126 @@
+                                      stdout=tempfd)
+
+                self.assertTrue(ret['rc'] == 0)
+                for line in tempfd:
```
Consider using ret['out'] instead of writing the output to a file and reading the file back in.

agrawalravi90 replied:
ret['out'] doesn't return the 'pristine' output of the command, it returns a list of the contents, and because I'm looking for something as subtle as whitespaces in the output, I wanted to look at the pristine, un-process output. That's why I think it's better to just write it out to a file and read it back in. What do you think?

bhroam replied:
True, whitespace is rather subtle.  It's easy for something to process it away without thinking much of it.  Writing out to a file is probably best.

bhroam commented on test/tests/functional/pbs_qmgr.py:
```diff
@@ -0,0 +1,126 @@
+            # Delete the vnode created
+            if vndef_file is not None:
+                self.mom.delete_vnodes()
+                self.server.manager(MGR_CMD_DELETE, VNODE, id=nodename)
```
This will happen in revert_to_defaults().  You don't need to do it.

agrawalravi90 replied:
True, but revert_to_defaults() is executed only during setUp(), this means that if one just ran a single test case, whatever setUp() it had created will linger on until you run another test case.

bhroam replied:
setUp() is called before every test.  No matter how many tests you run, revert_to_defaults() will be run.  The only thing this will do is delete it prior to leaving the final test.  I'd rather you leave it to revert_to_defaults().  It's the standard place this sort of thing is done in PTL.

agrawalravi90 commented:
Thanks for the review Bhroam! I've addressed your comments, please provide further feedback.

bhroam commented:
I sign off.  Go ahead and squash and rebase for the final merge.

agrawalravi90 commented:
Thanks bhroam, I've squashed the commits.
